docs: add the release process (#176)
maxGraph/maxgraph: packages/docs/release.md

# Release how-to

This page explains the steps needed to release a new version of maxgraph@core.

> Currently, the release is entirely manual. Automation will come later based on the information provided here.

Reviewer: ```suggestion
Currently, the release process is completely manual. Automation will come later based on the information provided here.
```

Reviewer: Some changes proposed by Deepl.
You are free to add them or not 😉

## Preparation

> Decide the new version depending on the types of changes:

Reviewer: ```suggestion
Decide on the new version depending on the type of changes:
```

- Follow [semver](https://semver.org/)

> - Until we release the first major version, bump the minor version when the release includes new features or breaking changes

Reviewer: ```suggestion
- Until we release the first major version, bump the minor version if the release contains new features or breaking changes.
```

> Review the milestone related to the new release. **Note:** we always put issues related to a version in a Milestone whose

Reviewer: ```suggestion
Check the milestone associated with the new release. **Note:** We always put issues related to a version into a Milestone whose
```

name matches the version.

> - Ensure the name of the milestone used for the new release version matches the name of the version that has is going to

Reviewer: ```suggestion
- Make sure that the name of the milestone used for the new release version matches the name of the version being
```

> be released. Renamed it if needed.

Reviewer: ```suggestion
released. Rename it if necessary.
```

> - Check that all issues related to the upcoming release are attached to the milestone. In particular, check the issues with

Reviewer: ```suggestion
- Verify that all issues related to the upcoming release are attached to the milestone. In particular, check the issues that
```

> [no milestone](https://github.com/maxGraph/maxGraph/issues?q=is%3Aissue+is%3Aclosed+no%3Amilestone).

Reviewer: ```suggestion
[do not have a milestone](https://github.com/maxGraph/maxGraph/issues?q=is%3Aissue+is%3Aclosed+no%3Amilestone).
```

> - Clean this opened milestone if some issues are still opened (move them to a new one or discard milestone from them)

Reviewer: ```suggestion
- Clean up this open milestone if some issues are still open (move them to a new milestone or discard the milestone from them).
```

> - Close the milestone

Reviewer: ```suggestion
- Close the milestone.
```

Changes in the source code
- Update the version in [package.json](../core/package.json) and the `VERSION` constant in the [Client](../core/src/Client.ts) file.

> - Update the [CHANGELOG](../../CHANGELOG.md) to list the major changes included in the new version. Be general and add a

Reviewer: ```suggestion
- Update the [CHANGELOG](../../CHANGELOG.md) to list the major changes included in the new version. Be generic and add a
```

> link to the future GitHub release that will include detailed release notes like in the following.

Reviewer: ```suggestion
link to the future GitHub release that will contain detailed release notes, as shown below.
```

```markdown

> For more details, see the [0.1.0 Changelog](https://github.com/maxGraph/maxGraph/releases/tag/v0.1.0) in the GitHub releases page.

Reviewer: ```suggestion
For more details, see the [0.1.0 Changelog](https://github.com/maxGraph/maxGraph/releases/tag/v0.1.0) on the GitHub release page.
```

```

> - Do a single commit including the changes described above.

Reviewer: ```suggestion
- Make a single commit that includes the changes described above.
```

> Create a git tag, prefixing the version by `v`. For instance, if the version is 0.2.0, run

Reviewer: ```suggestion
Create a git tag, prefixing the version with a `v`. For example, if the version is 0.2.0, run
```

```
git tag v0.2.0
git push origin v0.2.0
```

## Publish the npm package

- Checkout the tag that has just been created
- From packages/core, run `npm publish`

## Create the GitHub release

> **Example**: use the [maxGraph 0.1.0 release](https://github.com/maxGraph/maxGraph/releases/tag/v0.1.0) for inspiration

Reviewer: ```suggestion
**Example**: Use the [maxGraph 0.1.0 release](https://github.com/maxGraph/maxGraph/releases/tag/v0.1.0) for inspiration
```

Create a [new draft release](https://github.com/maxGraph/maxGraph/releases/)
- name: use the version that has just been published
- tag: use the tag create before
- save it as a draft

> Generate the list of the important changes by using the [automatically generated release notes](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes).

Reviewer: ```suggestion
Generate the list of the major changes by using the [automatically generated release notes](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes).
```

> It relies on the labels of the merged Pull Requests included in this version and the [GitHub release configuration](../../.github/release.yml).

Reviewer: ```suggestion
It is based on the labels of the merged Pull Requests included in this release and the [GitHub release configuration](../../.github/release.yml).
```

> If the list is not accurate (for instance, an element is not placed in the right category), update the label(s) or the related

Reviewer: ```suggestion
If the list is incorrect (for example, an item is not in the correct category), update the label(s) or the associated
```

Pull Request and regenerate the list.

On top of the auto-generated list, add a few words to highlight important changes. In particular, list **_breaking changes_**.

Also add links to the
- npm package
- GitHub milestone
- related paragraph in the Changelog file

Before you publish the release, make sure that a discussion will be created in the `Announces` category when the release
is published.

Publish the release.

Review the newly created discussion in the [Announces](https://github.com/maxGraph/maxGraph/discussions/categories/announces) category:
- adjust the title
- pin the discussion and unpin the previous release announce

> - see the [maxGraph 0.1.0 release announce](https://github.com/maxGraph/maxGraph/discussions/147) for instance.

Reviewer: ```suggestion
- see for example the [maxGraph 0.1.0 release announce](https://github.com/maxGraph/maxGraph/discussions/147).
```

## Update the integration examples repository

Once done, create a Pull Request in the [integration examples repository](https://github.com/maxGraph/maxgraph-integration-examples)
to use the new release. Use [PR #28](https://github.com/maxGraph/maxgraph-integration-examples/pull/28) as an example.

> **Note:** the examples will be updated automatically in the future, see the [issue #27](https://github.com/maxGraph/maxgraph-integration-examples/issues/27) for more details.

Reviewer: ```suggestion
**Note:** The examples will be updated automatically in the future, see the [issue #27](https://github.com/maxGraph/maxgraph-integration-examples/issues/27) for more details.
```
